### ISSUE TITLE
Hi-MD support?

### ISSUE THREAD

**Issue by benguild** [cybercase/webminidisc#117]
I'm curious about Hi-MD support, and if it's potentially to be supported at some point ever or if there are currently obstacles that make this unrealistic for the project's lifespan.

Hi-MD provides filesystem access, but according to `linux-minidisc` it seems like there's a special SCSI interface for some other features: https://wiki.physik.fu-berlin.de/linux-minidisc/himddiskformat

I'm mainly interested in organizing music into groups, but also converting some tracks to ATRAC3Plus would be nice. Doing this manually or in a Windows VM isn't ideal but it works! Would prefer something more modern if it's potentially roadmappable, but as mentioned, I'm curious about blockers.

**Comment by asivery** [comment]
@benguild I am currently working on adding HiMD support. It might not be possible to add AT3+ encoding (I am still not sure), but metadata editing works already :)

**Comment by GrayzerGray** [comment]
Seeet. Thank you for your effortsSent from my iPhoneOn 25 Dec 2022, at 14:57, asivery ***@***.***> wrote:﻿
@benguild I am currently working on adding HiMD support. It might not be possible to add AT3+ encoding (I am still not sure), but metadata editing works already :)

**Comment by benguild** [comment]
> @benguild I am currently working on adding HiMD support. It might not be possible to add AT3+ encoding (I am still not sure), but metadata editing works already :)

Great! Can't wait to try it. — AT3+ support would be great, but I'm not sure if it's supported by any tools.

Currently I use a Windows VM for managing things, and it'd be nice to have something more modern.

**Comment by asivery** [comment]
@benguild I got all HiMD functions working - AT3+ / AT3 / MP3 upload included. I will release the new version in the next few days.

**Comment by benguild** [comment]
oh hell yes. can't wait to try it out! I need to update some tracks on a disc so it's perfect timing.

**Comment by Invictaz** [comment]
This is mega cool. I hope these patches get ported to Platinum-MD

**Comment by Invictaz** [comment]
@benguild I asked for it in 2019. Wow.
It only took four years

**Comment by asivery** [comment]
@Invictaz It wasn't as simple back then. Getting it working in Chrome requires temporarily opening the normal MD mode, then exploiting the firmware and switching back. It's impossible without that, because WebUSB doesn't let you connect to mass storage device.
To get around that, I load up an exploit and patch the USB descriptor, so that the OS misidentifies the HiMD device.

**Comment by benguild** [comment]
> exploiting the firmware and switching back

is that going to work on Mac? 👀
are you exploiting the Hi-MD device, or the host system?

**Comment by asivery** [comment]
@benguild I'm exploiting the device, and I have confirmed it works on Mac, Windows and Linux

**Comment by benguild** [comment]
> @benguild I'm exploiting the device, and I have confirmed it works on Mac, Windows and Linux

that's hilarious, what an interesting discovery. is it a permanent exploit or just temporary? how does it work? very interesting!

**Comment by asivery** [comment]
@benguild It's all temporary, when you unplug power from the device, or the device goes to sleep, it reverts back. Feel free to look through a repo called 'netmd-exploits' on my Github, if you'd like to learn more about its inner workings.

**Comment by benguild** [comment]
> @benguild It's all temporary, when you unplug power from the device, or the device goes to sleep, it reverts back. Feel free to look through a repo called 'netmd-exploits' on my Github, if you'd like to learn more about its inner workings.

nice! I'm checking it out. Is there a way to force the player into charging but not data mode until it's rebooted? that'd be great for playback while charging from a PC/Mac.

**Comment by asivery** [comment]
@benguild I might be able to do it using exploits, for now it's not possible

**Comment by benguild** [comment]
> @benguild I might be able to do it using exploits, for now it's not possible

ah okay. was just wondering! right now I often do playback using a wall charger to avoid cycling the battery. (I just leave the battery out)

would be great to be able to use my Mac to run the player instead without it kicking over to data. I tried a USB-C data filter but that confused the player for some reason.

**Comment by Invictaz** [comment]
@asivery are Hi-MD transfers to computer (uploads) encrypted like in Sonicstage? Or decrypted? And how about downloads (from PC to device) ?

You mentioned AT AT3+ and MP3 but there is also lineair PCM.

**Comment by asivery** [comment]
@Invictaz At the time when I was typing that message, PCM wasn't working. Now it is. The uploads and downloads indeed depend on SonicStage's keys, but they work on unencrypted files. All the encryption happens in the app and on the devices.

**Comment by Invictaz** [comment]
But how to get those keys? You need SonicStage installed? I wonder why nobody reverse engineered the Sony system or cracked the 3DES encryption. It's considered inadequate by today's security standards.

**Comment by asivery** [comment]
@Invictaz They are hardcoded in the app

**Comment by Invictaz** [comment]
So there is a masterkey?

**Comment by asivery** [comment]
Yes - here it is: https://github.com/thp/linux-minidisc/blob/minidisc-ffwd/libhimd/encryption.c#L131

**Comment by asivery** [comment]
All other keys are generated on upload, then either XORed with this one, or DES encrypted / decrypted with it

**Comment by Invictaz** [comment]
What about decrypting older .OMA files that you lost the encryption keys from?

**Comment by asivery** [comment]
That's still impossible

**Comment by Invictaz** [comment]
Where are the keys stored on Sonicstage? Somewhere in %appdata% ?

**Comment by asivery** [comment]
Not sure, but here:
> why nobody reverse engineered the Sony system or cracked the 3DES encryption

you're wrong. I managed to add support for HiMD only thanks to the original linux-minidisc developers' notes, and by contacting them on the MD wiki discord. Nobody before re-implemented the DRM, but it had been reverse engineered long ago.

**Comment by Invictaz** [comment]
I am not wrong. You admitted that older .OMA tracks encrypted with 3DES are still not able to be cracked. That is something different than implementing DRM. I know of the efforts of the encryption algorythms and the QHiMDtransfer software did transfer a lot of Hi-MD functions already years ago.

**Comment by asivery** [comment]
Ah, I misunderstood you then, sorry. Indeed the 3DES keeping the tracks encrypted has not been cracked yet.

**Comment by Invictaz** [comment]
No worries you did a great job.

The thing is that we need to solve this matrix

![Screenshot_20230408_031043_Firefox Nightly](https://user-images.githubusercontent.com/20338545/230696929-5f0afd22-c3ce-4b89-ae57-b4bbbf7fa39c.jpg)

If we have filled in the blanks the linux-minidisc project is complete. I know they forked their project in a new branch called linux-minidisc-ffwd and are backporting or cherry picking fixes which are coded in their main repo. But that's very very slow.

The Sony system how they generate the keys is also not understood. The .dll for AT3 encoding is from my understanding closed source and never reverse engineered. 

Some of the DRM comes from OpenMG which is still available from Sony

https://www.sony-latin.com/en/electronics/support/downloads/W0001399

**Comment by asivery** [comment]
Per-track key generation is simple. Also, we solved the problem of a good quality ATRAC3 encoder with the help of the remote encoder. 
This is how the table looks like with this new version of Web Minidisc:
![Screenshot_20230408_033310](https://user-images.githubusercontent.com/82324156/230697560-da71f955-2fa2-4f5e-8141-d254e2cbafce.png)

Edit:
NetMD audio upload via USB works only on Sony portables, and I did omit the MZ-RH1 from having full support for all the ATRAC himd formats on purpose - it's one of the 2 unsupported devices - the MZ-RH1 and the MZ-DH10P.

**Comment by Invictaz** [comment]
The RH1 that is technically superior and should have all formats up and down? Why is it unsupported?

**Comment by asivery** [comment]
It's unsupported because my HiMD mode depends on firmware exploits. They patched the simple exploits in that device (they actually changed the whole main chip), and even though it has been exploited recently using another attack vector, I am not sure weather or not it is safe to include it in the main version of Web Minidisc. The MZ-RH1 is a very rare and expensive device - I don't want to brick any of them.
Edit: As far as the camera goes - that one is unsupported because I don't own one, and I don't even know what chips and firmware it uses

**Comment by benguild** [comment]
> The MZ-RH1 is a very rare and expensive device - I don't want to brick any of them.

womp womp. I have this player :(

**Comment by asivery** [comment]
@benguild I own an RH1 - I'll try to add support for it, but it might not work out in the end... In any case, on Linux it's possible to get the full HiMD mode working without any firmware exploits (just with ElectronWMD and not WMD in Chrome), so that's always a way out...

**Comment by benguild** [comment]
> @benguild I own an RH1 - I'll try to add support for it, but it might not work out in the end... In any case, on Linux it's possible to get the full HiMD mode working without any firmware exploits (just with ElectronWMD and not WMD in Chrome), so that's always a way out...

I might be interested in that if I can spin up a VM and burn tracks more quickly than with Windows. The Sony software is awful and tends to lock up when I virtualize a Windows system on an ARM machine just to use it.

**Comment by asivery** [comment]
The problem with the MZ-RH1 is:
The normal NetMD / HiMD units have (among others) these commands:
- Open / close mapped / eeprom memory for reading / writing / both
- Write previously open memory
- Read previously open memory
- Write DRAM (anti-shock memory)
- Read DRAM (anti-shock memory)

The MZ-RH1 does away with the first 3.
@Sir68k managed to reproduce the behavior of the read / write memory commands using the DRAM ones, but it works only for some sections of the device's memory map (You can access the DRAM and peripheral segments, not the device RAM or ROM). In theory, I only need access to the peripheral segment to get the required firmware patches running, but I haven't tested it.

**Comment by Invictaz** [comment]
First I am not a fan of doing exploits to please a Chrome environment for a device that should do it natively. That's why I bought it. And yes I understand that the sending of unencrypted tracks is different from encrypted ones using SonicStage. The whole pesky encryption is a pain in the eye. Exploiting a very rare, very expensive and sadly fragile device worth also emotional value to me is troublesome.

I have seen the repository with exploits and the chipset of the RH1 needs to be investigated more. The RH1 had only 1 firmware v1.0 and I wonder why there were no updates. There is a factory mode but if you use it wrong you end up with a laser which does not work anymore or a brick.

**Comment by asivery** [comment]
@Invictaz The whole problem with these HiMD devices (and the reason why no one wrote a program that would upload ATRAC3 / ATRAC3+ / PCM) is that the devices need the disc to be signed - there's an integrity check value, that's generated from some files on the disc, and later gets written to a region outside of the HiMD filesystem using Sony vendor SCSI commands.

Linux lets me disconnect the kernel mass storage driver, even after it had already been loaded. That is not the case with Windows (to use devices with WebUSB, you need to change the driver to a libusb-type one, with something like Zadig - you most likely could get it to work with ElectronWMD on there, but you'd lose the ability to access the files directly, since Zadig would override the base mass storage driver, at least from what I understand). I don't know how that would work on Mac. When I tried to disconnect the kernel driver on there, nodejs got terminated by the OS. Granted, all my Mac testing was done on Catalina, but I'm not sure the added an ability to detach an already loaded kernel driver.

The only thing my exploit does, is it rewrites the USB VID and PID pair, and changes the interface class from 0x08 subclass 0x05 (mass storage device using SCSI-over-ATAPI) to 0xFF subclass 0x00(vendor specific). This makes it possible to:
a) Use SonicStage alongside Web Minidisc Pro - the SonicStage driver looks for a device with a different VID/PID
b) Use the device within Chrome (WebUSB doesn't let you attach to devices which declare themselves as mass storage)
c) Let you still see the devices' filesystem, after the exploit clears itself

**Comment by Invictaz** [comment]
Having the Zadig swap already has its constraints but I prefer changing a pc based driver over an exploit. 

Regarding "the devices need the disc to be signed - there's an integrity check value, that's generated from some files on the disc, and later gets written to a region outside of the HiMD filesystem using Sony vendor SCSI commands."

More protection problems... sigh.

If you could document the vendor specific SCSI commands we would be further along the way.

**Comment by asivery** [comment]
@Invictaz I still need to do tests on how Windows behaves - everything I've said in this thread are purely just my expectations. I am developing all of this on Linux, because it's just a lot easier, and don't really have a Windows PC on hand to test everything on (VMs have problems with the device switching from NetMD mode to HiMD mode).

I'll do my best to document the SCSI commands, but most of that has already been documented over at the linux-minidisc wiki.

Edit: The only thing I tested on windows is the exploit method with my MZ-NH700, and it worked.

**Comment by Invictaz** [comment]
If we would have an RH1 donor board it would be much easier

**Comment by asivery** [comment]
Why? We already have the MZ-RH1's firmware

**Comment by GrayzerGray** [comment]
Superb, can’t wait to get trying that. Great efforts providing this feature, I know it’s been alluding you for a while. Well done and thank youGraySent from my iPhoneOn 8 Apr 2023, at 00:16, asivery ***@***.***> wrote:﻿
@Invictaz It wasn't as simple back then. Getting it working in Chrome requires temporarily opening the normal MD mode, then exploiting the firmware and switching back. It's impossible without that, because WebUSB doesn't let you connect to mass storage device.
To get around that, I load up an exploit and patch the USB descriptor, so that the OS misidentifies the HiMD device.

**Comment by GrayzerGray** [comment]
Thankfully I have the RH1 in mint other than the famous OLED failure (Not brave enough to do the fix yet). And also have an MZ-RH910 so all good. If anyone is on the hunt for an RH1 boxed NOS will set you back £1000. Battery most likely shot and no guarantee the OLEDs will not be good. Mine still cost £400 in the condition it was in and factored in an OEM battery from AliExpress. But thank you again for getting Hi-MD functionality out of itSent from my iPhoneOn 8 Apr 2023, at 02:47, Ben Guild ***@***.***> wrote:﻿

The MZ-RH1 is a very rare and expensive device - I don't want to brick any of them.

womp womp. I have this player :(

**Comment by asivery** [comment]
@benguild I got the full HiMD mode running on the MZ-RH1

**Comment by Invictaz** [comment]
@asivery If we would have a donor board we could test without sacrificing a good working MZ-RH1.

How did you manage to get it fully working?

**Comment by asivery** [comment]
@Invictaz Recently, @Sir68k managed to dump the firmware from his MZ-RH1. I just used the same exploit to dump mine, then reverse engineered where the firmware keeps the values I need to change, and patched them as I would in a normal HiMD device.

**Comment by Invictaz** [comment]
Great work. What has changed since last week?

**Comment by asivery** [comment]
I just thought that if I can implement that functionality, I should. There will be an appropriate warning in the app before running the exploit code. If someone decides to continue, then bricks their device, that's on them.

**Comment by benguild** [comment]
> I just thought that if I can implement that functionality, I should. There will be an appropriate warning in the app before running the exploit code. If someone decides to continue, then bricks their device, that's on them.

Is there a non-exploity way to get this running on Linux? I might just spin up a VM versus brick risk. The Windows NetMD software is terrible, slow, and unstable and if there's any modern option I imagine it has to be better.

My device is from Japan so the firmware is likely different as it displays characters in Japanese on some screens.

**Comment by asivery** [comment]
You can change the language - the firmware is most likely the same. As for the non-exploity way - yes. If you're running Linux, the same rule applies as with other HiMD devices. You can just access them normally.

**Comment by benguild** [comment]
> You can change the language - the firmware is most likely the same. As for the non-exploity way - yes. If you're running Linux, the same rule applies as with other HiMD devices. You can just access them normally.

How do you manage discs on Linux? Do you use a UI or the command line? I have a SMB share on Mac with folders for playlists that I access from the Windows VM, so it's pretty straightforward... but then the NetMD software is slow and annoying.

So, assuming I have access to the media to transfer, what's the best way to do this? Is there a guide or README put together? I'm starting from zero.

**Comment by asivery** [comment]
@benguild You can simply mount the SMB share on Linux - there's no problem with that. Since I use Linux as my main OS, I don't have the problem of mounting shares from the host. Also, it's possible that MacOS does support that non-exploit mode that Linux supports. I only have access to a MacBook running Catalina, and that doesn't work. Maybe it works with a newer OS? I'm not sure.

Edit: There's no READMEs written as of yet, because I haven't released the software for now.

**Comment by GrayzerGray** [comment]
Seeet well done and congrats ngratsSent from my iPhoneOn 16 Apr 2023, at 18:48, asivery ***@***.***> wrote:﻿
@benguild I got the full HiMD mode running on the MZ-RH1

**Comment by asivery** [comment]
Apologies for the delay. HiMD mode is pretty much 100% done now (including MZ-RH1 support 😄  ), except for some UI glitches which I'll take care of soon.
@Invictaz, @benguild, @GrayzerGray If anyone of you would like to test it before I update web.minidisc.wiki, it would be greatly appreciated.
The current snapshot is available at: https://testing.minidisc.wiki/webminidisc-testing/
And the exploit source code: https://github.com/asivery/netmd-exploits/blob/master/src/exploits/himd-usb-class-override.ts

Here's the procedure you need to follow to get to HiMD mode.
First, some info about both the HiMD mode in Web Minidisc and the HiMD mode on the devices themselves.
The player will always switch to accommodate the disc currently inserted. A regular MD with at least one track will force the HiMD device into MD mode, and a HiMD disc with at least one track will force the device to HiMD mode.
Blank discs are an exception to that rule, they are treated as format-less, and will be formatted by the device to the mode it's currently in.

To enter HiMD mode in WebMinidisc, you need to:
1. Connect the HiMD device to the computer, while it's in MD mode (either with a regular MD inside, which has at least one track, or just set to MD disc mode and without a disc)
2. Connect it to WebMinidisc, as you would do if you wanted to use normal NetMD mode.
3. Open the '3 dots' menu, select settings, then 'enable homebrew mode shortcuts', then close the settings (this step needs to be done only once)
4. Open the '3 dots' menu, select homebrew shortcuts, then 'switch to unrestricted HiMD mode'
5. At this point, you'll see Web Minidisc's loading screen, then a 'Loaded.' dialog will appear. After closing the dialog you'll be put back at the 'connect' screen.
--- At this point the exploit is already loaded onto the device and it will stay loaded until the device goes to sleep, unplugging the USB cable will not erase the exploit ---
7. Insert a HiMD disc (either a HiMD formatted MD with at least one track, or a blank one with the device set to HiMD mode)
8. (Windows only) Install Zadig again for this new HiMD device. It should have 0x5341 as the VID.
9. Click the downwards-pointing arrow next to the connect button, then select HiMD (Unrestricted). A WebUSB dialog will appear. Select your device, then wait.
10. Done. You can now upload AT3/3+/PCM/MP3 audio

Known bugs (and temporary mitigations) - they will be fixed in the final version:
- Out of the box, the selected encoder is 'atracdenc', which doesn't let you convert ATRAC3+, but the UI will still have AT3+ selected as the default. Fix: Please select the remote encoder in the settings.
- Full width support is broken - the concept of full width titles doesn't exist in HiMD. Fix: If you want to name a track using Japanese characters, you'll have to do it after the upload.

Please let me know if you encounter any bugs, other than the ones I've mentioned here.

Edit: Add step 8

**Comment by asivery** [comment]
If the "Switch to unrestricted HiMD button" is greyed out, that means your firmware version is not supported yet. In that case, please open the menu again, click "Enter homebrew mode". Once that mode loads open the menu and select "Read Firmware", then please share the files it downloads here and I'll add support for it.

**Comment by Invictaz** [comment]
Good work but I'm still not sure I want to risk a brick by doing the exploit on a insanely expensive device.

**Comment by asivery** [comment]
@Invictaz Of course, if you're not feeling comfortable doing it, I fully respect that.
I'll implement a way to run it without exploits in ElectronWMD, but before I do that, I needed a working base, and that's what this version is.
A brick should be pretty much impossible (at least using this exact exploit), but I'm not taking any responsibility for it if it does brick.

**Comment by Invictaz** [comment]
If there was a Sony tool to reflash the firmware or do an emergency recovery hard reset or such, I would consider it. But to my knowledge there isnt. Sony never released any flashable firmware for the RH1. 

ElectronWMD is quite heavy on resources?

**Comment by asivery** [comment]
HiMD devices have a hidden mode that allows for reflashing, but we haven't found a way to trigger it. The FLASH ROM is usually write-protected, and you need to issue special commands in order to unlock it. The only thing that could be a potential "brick" would be erasing the EEPROM contents (encryption keys, laser power config, etc..., but I'm not touching any EEPROM code with this exploit). And that shouldn't happen, because before exploiting anything, `netmd-exploits` checks the firmware version of your device, and applies modifications only if it had been [whitelisted](https://github.com/asivery/netmd-exploits/blob/1553c1ff8c275593096edc92dfd276ebc4bb60f4/src/compatibility.ts#L29).

Edit: ElectronWMD is a typical electron app, it shouldn't be much heavier than f.ex.  Discord.

**Comment by Invictaz** [comment]
As with all homebrew and exploits I wait until it had a bit more time in the oven. 

Also it is sad that the RH1 is so locked down as Sony marketed it as the final and unrestricted MD device. 

I applaud your work though. It is a very very good achievement. And hopefully you will be able to do the Hi-MD decks as well. Or don't they use USB?

**Comment by asivery** [comment]
Thanks! Some bugs will probably rear their ugly heads when I release this version, but I'll try to resolve them as they show up.
> Also it is sad that the RH1 is so locked down as Sony marketed it as the final and unrestricted MD device.

The RH1 is not locked down, chrome is. If it wasn't for chrome's restrictions, none of this would be needed.

> And hopefully you will be able to do the Hi-MD decks as well

Exploits probably won't work for them, but regular ElectronWMD should work, as long as they have USB.

**Comment by Invictaz** [comment]
What about Edge Chromium? It's sad that Firefox is against webUSB

**Comment by asivery** [comment]
The browser doesn't make a difference. WebUSB spec specifies that browsers must not be able to connect to devices which report as being able to run as mass storage (USB Interface Class = 0x08). That's the value the HiMD machines report. My exploit essentially locates that value (0x08) in the device's RAM, then replaces it with 0xFF (Generic / Proprietary device). This bypasses that limitation.

**Comment by benguild** [comment]
Is it more realistic to modify Chrome rather than exploit the player's firmware?

**Comment by asivery** [comment]
@benguild Nope, not really. If you're willing to modify the browser, it's better to just use the Electron version.

**Comment by benguild** [comment]
﻿Is there a disclaimer to folks in the UI or CLI prior to running the exploit? I doubt anyone would want to risk bricking an expensive device unless it was already a near or total brick.

I appreciate all of the hard work but the players are harder to access than NetMD is on a Windows VM, so it’s a bit of a tough sell honestly.

**Comment by asivery** [comment]
It's not available from any CLI, and unlocking any homebrew features has disclaimers in the UI. I think I might have oversold this exploit's ability to brick devices. It has been tested multiple times by a few people already (including me on my RH1), and no bricks occurred.

I understand that this might be not as convenient to access as NetMD is, but unfortunately these are just the limitations of Chrome. Again, I will try to make it so that ElectronWMD doesn't need the exploit, but ElectronWMD's release will happen after this version of Web Minidisc. Thanks for the feedback 😃 

Edit: This exploit is just as likely to brick as the ATRAC download exploits on normal NetMD devices, and no bricks occurred to anyone using that in the span of ~1 year since it had been released.

**Comment by asivery** [comment]
@benguild I got full HiMD working without exploits in ElectronWMD. Unfortunately I've run into some trouble related to getting it to work on MacOS. It might not be possible to get it working without root access there. On Linux I can define UDEV rules once, that authorize a given user or group to have full access to a device. Do you know if it's possible to do something like that on MacOS? I'm asking because from what I remember you're on Mac, and I'd rather not run an electron app as root.

**Comment by benguild** [comment]
I'm not sure. Generally apps with specific entitlements can perform things outside of the sandbox. https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_device_usb

If something needs root access or something specific it may be better to ship a separate process that just does that and call upon it.

**Comment by asivery** [comment]
@benguild I managed to add HiMD exploit-free support to MacOS in ElectronWMD.
I'm looking for testers now, before I release this version. If you'd be willing to see if it works correctly on your hardware it'd be great!
You can either compile it from my repo, or download it here ([arm64](https://testing.minidisc.wiki/electronwmd/electronwmd-0.3.0-1.4.0-mac_arm64.dmg) / [x86_64](https://testing.minidisc.wiki/electronwmd/electronwmd-0.3.0-1.4.0-mac_x64.dmg)).

To connect to a device, it's not necessary to go through the disc swapping and exploit procedures using this version - it's as simple as clicking "Connect (HiMD Full)" in the UI when the device is visible to the OS as mass storage.

There's one problem though - you need to wait for MacOS to stop writing its proprietary files to the medium. If you click that button when MacOS is still using it, both the HiMD device and ElectronWMD will crash.

Please let me know if you encounter any issues with it, if you decide to test it. :)

**Comment by Invictaz** [comment]
@asivery Would it be possible to trigger a warning pop up to show "wait before unplug"?

**Comment by asivery** [comment]
@Invictaz Do you mean something like a "Please make sure the device is not in use, then press OK" dialog?

**Comment by Invictaz** [comment]
A bit more specific since the user does not know if it is the device or Mac writing its specifics to the device

**Comment by asivery** [comment]
Since this feature has been added in both WMDPro v1.4.0 and EWMD v0.3.1-1.4.2, I'm closing this issue. HiMD mode is now fully supported in both exploit and exploit-free mode on all platforms.
